### PR TITLE
Add lmbench scripts

### PR DESCRIPTION
**Pull request by sdww0**
This PR adds lmbench scripts to show the performance of Linux and Asterinas. Additionally, this PR adds several benchmark building scripts(LMbench, Unixbench, Iozone) in the Dockerfile, and it needs to update the docker image.

### CODE REVIEW

File: .github/workflows/benchmark_asterinas.yml [hunk 1]
```diff
@@ -11,7 +11,11 @@ jobs:
     runs-on: self-hosted
     strategy:
       matrix:
-        benchmark: [sysbench-cpu, sysbench-thread, getpid]
+        benchmark: |
+          [
+          sysbench-cpu, sysbench-thread, getpid,
+          lmbench-getpid, lmbench-fork, lmbench-signal
```
Comment (tatetian) [inline .github/workflows/benchmark_asterinas.yml:17]: What about `lmbench-read` and `lmbench-write`? And are they any other bench programs in lmbench that are missing in this list? If any, why?

**Review by tatetian — COMMENTED** [review]
Could you provide a link to the HTML output of the benchmark results so that I can preview the visualized output?

**Comment by tatetian** [comment]
@sdww0 I suggest splitting this PR into two halves, submitting the changes to the Docker image as the first PR.

**Comment by cchanging** [comment]
I find that the results of the lmbench fork test can vary significantly (./lat_proc -P 1 fork), sometimes is 400 micro secs, and sometimes may even reach 8000 micro secs. Have you encountered similar issues @sdww0 ? I'm not sure if this is due to some bugs; if not, we might need to increase the number of repetitions to mitigate these discrepancies.

**Comment by sdww0** [comment]
> I find that the results of the lmbench fork test can vary significantly (./lat_proc -P 1 fork), sometimes is 400 micro secs, and sometimes may even reach 8000 micro secs. Have you encountered similar issues @sdww0 ? I'm not sure if this is due to some bugs; if not, we might need to increase the number of repetitions to mitigate these discrepancies.

I haven't encountered this issue. Have you been running lmbench without Docker? I'm not using Docker and have been consistently around 300ms in my tests.

File: test/benchmark/lmbench-fork/config.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "alert_threshold": "450%",
```
[the remark this answers is not in the document]
Comment (sdww0) [reply test/benchmark/lmbench-fork/config.json:2]: The latency of fork on Linux is 66ms compared to 297ms on Asterinas. Therefore, I set 450% as the alert threshold. 

This value should be adjusted once we've addressed the performance gap.

**Comment by cchanging** [comment]
>  Have you been running lmbench without Docker?

I run it with docker in main branch, only adding pselect6 syscall like this PR.

File: tools/docker/Dockerfile.jinja [hunk 3]
```diff
@@ -62,10 +68,36 @@ RUN ./configure --prefix=/usr/local/benchmark/iperf \
...
+    && mv bin/x86_64-linux-gnu /usr/local/benchmark/lmbench \
+    && rm /usr/local/benchmark/lmbench/*.o \
+    && rm /usr/local/benchmark/lmbench/*.a 
+
+# Build unixbench
+WORKDIR /root/byte-unixbench/UnixBench
+# Disable sse2
+RUN sed "s/-O3 -ffast-math/-O3 -ffast-math -mno-sse2/g" Makefile \
+    > Makefile.tmp \
+    && mv Makefile.tmp Makefile
```
[the remark this answers is not in the document]
Comment (sdww0) [reply tools/docker/Dockerfile.jinja:83]: Resolve in #1023

File: tools/docker/Dockerfile.jinja [hunk 4]
```diff
@@ -62,10 +68,36 @@ RUN ./configure --prefix=/usr/local/benchmark/iperf \
...
+RUN make -j && mv pgms /usr/local/benchmark/unixbench 
+
+# Build iozone
+WORKDIR /root/iozone3_506/src/current
+RUN CFLAGS=-static make linux-AMD64 \
+    && mkdir /usr/local/benchmark/iozone \
+    && cp iozone /usr/local/benchmark/iozone/
```
[the remark this answers is not in the document]
Comment (sdww0) [reply tools/docker/Dockerfile.jinja:90]: Resolve in #1023

File: .github/workflows/benchmark_asterinas.yml (hunk 1 above)
Comment (tatetian) [reply .github/workflows/benchmark_asterinas.yml:17]: I see. Then I suggest adding the two tests to the list, but comment them out with a TODO that details the reason why they are disabled for now.

**Comment by tatetian** [comment]
@sdww0 Could you provide a link to the HTML output of the benchmark results so that I can preview the visualized output?

File: test/benchmark/lmbench-fork/config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply test/benchmark/lmbench-fork/config.json:2]: No, the threshold is NOT about the gap between Linux and Asterinas, but the one between the current and the previous commit of Asterinas.

**Comment by sdww0** [comment]
> @sdww0 Could you provide a link to the HTML output of the benchmark results so that I can preview the visualized output?

I manually added the results according to the `asterinas/benchmark` repository. Results: https://github.com/sdww0/benchmark

File: test/benchmark/lmbench-fork/result_template.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+[
+    {
+        "name": "Average Fork latency on Linux",
```
Comment (tatetian) [inline test/benchmark/lmbench-fork/result_template.json:3]: Fork -> fork

File: test/benchmark/lmbench-fork/result_template.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+        "unit": "ms",
+        "value": 0,
+        "extra": "linux_avg"
+    },
+    {
+        "name": "Average Fork latency on Asterinas",
```
Comment (tatetian) [inline test/benchmark/lmbench-fork/result_template.json:9]: Fork -> fork

File: test/benchmark/lmbench-signal/result_template.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+[
+    {
+        "name": "Average Signal handler overhead on Linux",
```
Comment (tatetian) [inline test/benchmark/lmbench-signal/result_template.json:3]: Signal -> signal

File: test/benchmark/lmbench-signal/result_template.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+        "unit": "ms",
+        "value": 0,
+        "extra": "linux_avg"
+    },
+    {
+        "name": "Average Signal handler overhead on Asterinas",
+        "unit": "ms",
+        "value": 0,
+        "extra": "aster_avg"
```
Comment (tatetian) [inline test/benchmark/lmbench-signal/result_template.json:12]: Signal -> signal

File: test/benchmark/lmbench-signal/result_template.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+[
+    {
+        "name": "Average Signal handler overhead on Linux",
+        "unit": "ms",
+        "value": 0,
+        "extra": "linux_avg"
+    },
+    {
+        "name": "Average Signal handler overhead on Asterinas",
+        "unit": "ms",
```
Comment (tatetian) [inline test/benchmark/lmbench-signal/result_template.json:10]: Use "us" as the unit

**Review by tatetian — COMMENTED** [review]
Only some minor issues left. And please address the issues raised before.

File: test/benchmark/lmbench-signal/result_template.json (hunk 9 above)
Comment (sdww0) [reply test/benchmark/lmbench-signal/result_template.json:10]: > Use "us" as the unit

Using "us" as the unit requires multiplying the result by 1000. However, the current benchmarking design does not support this function. I suggest using the `value` field of the `result_template.json` to indicate the multiplier and this function should not be implemented in this PR.

**Comment by sdww0** [comment]
> > Have you been running lmbench without Docker?
> 
> I run it with docker in main branch, only adding pselect6 syscall like this PR.

It reminds me of a previous experience. I ran Asterinas with `Unixbench-fork` on a different Intel CPU. The two computers displayed completely different performance gaps compared to Linux, with one at 4x and the other at 10x. Although it's similar to what you mentioned, it seems that the causes might be different.

File: test/benchmark/lmbench-signal/result_template.json (hunk 9 above)
Comment (tatetian) [reply test/benchmark/lmbench-signal/result_template.json:10]: I see. Then let's keep using "ms".

**Comment by tatetian** [comment]
> > > Have you been running lmbench without Docker?
> > 
> > 
> > I run it with docker in main branch, only adding pselect6 syscall like this PR.
> 
> It reminds me of a previous experience. I ran Asterinas with `Unixbench-fork` on a different Intel CPU. The two computers displayed completely different performance gaps compared to Linux, with one at 4x and the other at 10x. Although it's similar to what you mentioned, it seems that the causes might be different.

So the benchmark results are stable on a fixed machine, but could vary significantly from machine to machine?

**Comment by tatetian** [comment]
@sdww0 Need to fix the conflict.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**